src/background.c

```diff
@@ -679,69 +643,92 @@ check_background_apps (void)
       const char *state_names[] = { "background", "running", "active" };
-      g_autofree char *handle = NULL;
+      gboolean is_new = FALSE;
 
       if (!flatpak_instance_is_running (instance))
         continue;
 
-      state = get_one_app_state (app_id, app_states);
-      g_debug ("App %s is %s", app_id, state_names[state]);
+      id = flatpak_instance_get_id (instance);
+      app_id = flatpak_instance_get_app (instance);
+      idata = g_hash_table_lookup (applications, id);
 
-      if (state != BACKGROUND)
+      if (!idata)
         {
-          remove_background_app (app_id);
-          continue;
+          is_new = TRUE;
+          idata = g_new0 (InstanceData, 1);
+          idata->instance = g_object_ref (instance);
+          g_hash_table_insert (applications, g_strdup (id), idata);
         }
 
-      /* If the app is not in the list of background apps
-       * yet, add it, but don't notify yet - this gives
-       * apps some leeway to get their window app. If it
-       * is still in the background next time around,
-       * we'll proceed to the next step.
-       */
-      if (!lookup_background_app (app_id, &handle))
-        {
-          add_background_app (app_id, NULL);
-          continue;
-        }
+      idata->stamp = stamp;
+      idata->state = get_one_app_state (app_id, app_states);
 
-      if (handle)
-        continue; /* already notified */
+      g_debug ("App %s is %s", app_id, state_names[idata->state]);
 
       permission = get_one_permission (app_id, perms);
-      if (permission == NO)
+
+      if (idata->permission != permission)
         {
-          pid_t pid = flatpak_instance_get_child_pid (instance);
-          g_debug ("Killing app %s (child pid %u)", app_id, pid);
-          kill (pid, SIGKILL);
+          /* Notify again if permissions change */
+          idata->permission = permission;
+          idata->notified = FALSE;
         }
-      else if (permission == ASK || permission == UNSET)
+
+      /* If the app is not in the list yet, add it,
+       * but don't notify yet - this gives apps some
+       * leeway to get their window app. If it is still
+       * in the background next time around, we'll proceed
+       * to the next step.
+       */
+      if (idata->state != BACKGROUND || idata->notified || is_new)
+        continue;
+
+      switch (idata->permission)
         {
-          send_notification (instance, permission);
+        case NO:
+          kill_instance (id);
+          idata->stamp = 0;
+          break;
+
+        case ASK:
+        case UNSET:
+          send_notification (idata);
+          break;
+
+        case YES:
+        default:
+          break;
         }
     }
+
+  remove_outdated_instances (stamp);
 }
 
 static gpointer
 background_monitor (gpointer data)
 {
+  applications = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                        g_free, instance_data_free);
   while (1)
     {
       check_background_apps ();
```

Comment:
You could pass applications here instead of relying on a global maybe?

Reply:
with the locking back, it is now more clearly a global